### PR TITLE
[ROCm] Enable MIOpen fused convolution relu

### PR DESCRIPTION
**Pull request by jeffdaily**
Adds MIOpen fused convolution relu for fp32 and contiguous memory format.  Adds fallbacks for conv + z + bias + relu, fp16, and channels last until MIOpen adds these features.

cc @sunway513 @jithunnair-amd @ROCmSupport @KyleCZH

### CODE REVIEW

File: test/jit/test_freezing.py [hunk 1]
```diff
@@ -2189,78 +2189,89 @@ def test_conv_to_mkldnn_no_mkldnn(self):
...
 
-    @unittest.skipIf(not TEST_CUDNN, "requires CUDNN")
+    @unittest.skipIf(not (TEST_CUDNN or TEST_WITH_ROCM), "requires CUDNN")
     def test_freeze_conv_relu_fusion(self):
-        conv_bias = [True, False]
-        conv_ops = [nn.Conv2d, nn.Conv3d]
-        add_z = [True, False]
-        use_tracing = [True, False]
-        for use_bias, conv, add_z, tracing in product(conv_bias, conv_ops, add_z, use_tracing):
+        with set_default_dtype(torch.float):
```
Comment (jeffdaily) [inline test/jit/test_freezing.py:2194]: For reviewers: The diff formatting is terrible here.  I added a context manager to change the default dtype (see _many_ similar examples elsewhere in this file), and indented the rest of the test.  I did the same in the next test.  As a result, this was done in two consecutive tests, and the diff was rendered in a way that obfuscates the change.

The changes to focus on:
- Calling new miopen routine if TEST_WITH_ROCM.
- Tests are indented inside context manager to temporarily set the default dtype.  MIOpen does not support double type convolutions.

**Comment by jithunnair-amd** [comment]
@ngimel please review

File: aten/src/ATen/native/miopen/Conv_miopen.cpp [hunk 2]
```diff
@@ -1449,6 +1463,219 @@ Tensor miopen_convolution_transpose(
...
+      groups,
+      benchmark,
+      false // deterministic
+  );
+
+  auto contig_output = self_or_new_memory_format(output, memory_format);
+
+  if (!output.is_same(contig_output)) {
+    contig_output.copy_(output);
+  }
```
Comment (ngimel) [inline aten/src/ATen/native/miopen/Conv_miopen.cpp:1568]: Can you call `output.contiguous(memory_format)` instead of this?

File: test/jit/test_freezing.py (hunk 1 above)
Comment (ngimel) [reply test/jit/test_freezing.py:2194]: Viewing with "hide whitespace" renders with only the relevant changes.

File: aten/src/ATen/native/miopen/Conv_miopen.cpp [hunk 3]
```diff
@@ -1449,6 +1463,219 @@ Tensor miopen_convolution_transpose(
...
+          && input.scalar_type() == at::kFloat
+          && input.ndimension() == 4) {
+
+    // FuseFrozenConvAddRelu performs some tensor shape checking
+    Tensor output_t = at::detail::empty_cuda(
+        conv_output_size(
+            input.sizes(), weight.sizes(), padding, stride, dilation),
+        input.options().memory_format(input.suggest_memory_format()));
+    if (output_t.numel() == 0) {
+      return output_t;
```
Comment (ngimel) [inline aten/src/ATen/native/miopen/Conv_miopen.cpp:1612]: I don't know if that's important for your usecases (that's inference probably?) but this would break the gradient chain, and weight and bias would not get gradients (weight.grad would be None)

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/miopen/Conv_miopen.cpp (hunk 3 above)
Comment (jeffdaily) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp:1612]: There was precedent in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/Conv_v8.cpp to return if `numel() == 0`.  Is that what you were referring to?

Comment (ngimel) [reply aten/src/ATen/native/miopen/Conv_miopen.cpp:1612]: Yeah there's precedent, we are not too good at preserving gradients through 0-element tensors.

**Review by malfet — APPROVED** [review]
